gradle/maven-publish.gradle

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2019 Benjamin Bader
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+apply plugin: "maven-publish"
+apply plugin: "signing"
+
+def isReleaseBuild() {
+    return !VERSION_NAME.contains("SNAPSHOT")
+}
+
+def getRepositoryUsername() {
+    return hasProperty('SONATYPE_NEXUS_USERNAME') ? SONATYPE_NEXUS_USERNAME : System.env.SONATYPE_NEXUS_USERNAME
+}
+
+def getRepositoryPassword() {
+    return hasProperty('SONATYPE_NEXUS_PASSWORD') ? SONATYPE_NEXUS_PASSWORD : System.env.SONATYPE_NEXUS_PASSWORD
+}
+
+def getRepositoryUrl() {
+    return isReleaseBuild()
+        ? "https://oss.sonatype.org/service/local/staging/deploy/maven2/"
+        : "https://oss.sonatype.org/content/repositories/snapshots/"
+}
+
+def hasDeploymentTask(Project project) {
+    def deploymentTasks = [
+        "publishPlugins",
+        "publish"
+    ]
+    deploymentTasks.any { project.gradle.taskGraph.hasTask(it) }
+}
+
+task sourceJar(type: Jar) {
+    archiveClassifier.set("sources")
+    from sourceSets.main.allSource
+}
+
+task javadocJar(type: Jar, dependsOn: dokka) {
```

Comment:
Do you need one for dokka? 


```
task docsJar(type: Jar, dependsOn: dokka) {
  group = 'Publications'
  description = 'Create jar of documentation.'
  archiveClassifier = 'docs'
  from dokka.outputDirectory
}
See: https://github.com/jaredsburrows/gradle-license-plugin/blob/master/gradle/publish.gradle#L35
```

Reply:
I don't think so?  This change uses Dokka for all documentation, replacing (or just ignoring) the Javadoc and Groovydoc jars we were generating before.  At the bottom of this task you can see that it's set up to contain only dokka output.  I might be missing something though, what do you think?

It's how I did it in Thrifty, so I just copypasta'd: https://github.com/microsoft/thrifty/blob/master/gradle/gradle-mvn-push.gradle#L108.

Reply:
Oh sorry, I read "javadoc" and thought of Java.